### PR TITLE
Add volume fee overrides

### PR DESCRIPTION
**Pull request by fafk**
# Description

Implements volume fee bucket overrides, allowing custom volume fees for groups of tokens. This enables us to have different fees for a chosen set of tokens, e.g. stable coins. 

# Changes

- [x] Added `--volume-fee-bucket-overrides` CLI argument
  - Format: `"factor:token1,token2,..."` (e.g., `"0:0xA0b86...,0x6B175...,0xdAC17..."`)
  - Multiple buckets separated by semicolons
- [x] Both buy and sell tokens must be in the same bucket for the override to apply
- [x] Adjusts quote (orderbook) and order (autopilot)
- [x] For buy token=sell token no fee applies 

## How to test

```
cargo test --package e2e --test e2e protocol_fee::local_node_volume_fee_overrides
```

### CODE REVIEW

File: crates/solver/src/liquidity_collector.rs [hunk 1]
```diff
@@ -185,10 +185,10 @@ mod test {
         };
 
         let source =
-            BackgroundInitLiquiditySource::new("fake", init, Duration::from_millis(10), None);
+            BackgroundInitLiquiditySource::new("fake_delayed_init", init, Duration::from_millis(10), None);
```
Comment (fafk) [inline crates/solver/src/liquidity_collector.rs]: Not related to the ret of the PR, just fixes flakiness of this test when run in parallel with other uni tests.

File: crates/autopilot/src/domain/fee/policy.rs [hunk 2]
```diff
@@ -84,13 +84,23 @@ impl PriceImprovement {
...
             boundary::OrderClass::Market => None,
             boundary::OrderClass::Liquidity => None,
-            boundary::OrderClass::Limit => Some(domain::fee::Policy::Volume {
-                factor: self.factor,
-            }),
+            boundary::OrderClass::Limit => {
+                // No volume fee for same-token trades
+                if order.data.buy_token == order.data.sell_token {
+                    return None;
+                }
```
Comment (MartinquaXD) [inline crates/autopilot/src/domain/fee/policy.rs]: Which fee gets applied is now spread across 3 places in the code:
1. base fee
2. fee override
3. this extra exemption

IMO there should be 1 function we can call that already provides the final fee policy taking every override and edge case into account.

File: crates/e2e/tests/e2e/protocol_fee.rs [hunk 3]
```diff
@@ -983,3 +993,384 @@ async fn volume_fee_buy_order_upcoming_future_test(web3: Web3) {
 
     assert_eq!(order.metadata.executed_fee, balance_after);
 }
+
+/// Volume fees can be overriden by defining "buckets" of tokens that have
+/// differnt vol fees than the default. If an order has both the buy and sell
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/protocol_fee.rs]:
```suggestion
/// different vol fees than the default. If an order has both the buy and sell
```

File: crates/e2e/tests/e2e/protocol_fee.rs [hunk 4]
```diff
@@ -983,3 +993,384 @@ async fn volume_fee_buy_order_upcoming_future_test(web3: Web3) {
...
+            .from(solver.address())
+            .send_and_watch()
+            .await
+            .unwrap();
+    }
+
+    // Configure fee policies:
+    // - Default volume fee: 1% (0.01)
+    // - Bucket 1 (2-token pair): USDC-DAI has 0.05% fee (checked first)
+    // - Bucket 2 (stablecoins): USDC, DAI, USDT have 0% fee (checked second)
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/protocol_fee.rs]: nit: these comments apply to the next code block

File: crates/e2e/tests/e2e/protocol_fee.rs [hunk 5]
```diff
@@ -983,3 +993,384 @@ async fn volume_fee_buy_order_upcoming_future_test(web3: Web3) {
...
+        .expect("Timeout waiting for the orders to trade");
+
+    // Verify fees
+    tracing::info!("Checking executions...");
+
+    // Get gas price from the node to verify network fee calculation
+    let gas_price_output = std::process::Command::new("cast")
+        .args(["gas-price", "--rpc-url", "http://localhost:8545"])
+        .output()
+        .expect("failed to get gas price");
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/protocol_fee.rs]: Why are you using `cast` to get the current gas price? This can be done using the `Web3` instance available in every e2e test.

File: crates/e2e/tests/e2e/protocol_fee.rs [hunk 6]
```diff
@@ -983,3 +993,384 @@ async fn volume_fee_buy_order_upcoming_future_test(web3: Web3) {
...
+        dai_usdt_fee,
+        usdc_weth_fee
+    );
+
+    // Sanity check: fees should be ordered by protocol fee percentage: 0% < 0.05% <
+    // 1%
+    assert!(
+        dai_usdt_fee < usdc_dai_fee && usdc_dai_fee < usdc_weth_fee,
+        "Fees should be ordered: DAI-USDT (0%) < USDC-DAI (0.05%) < USDC-WETH (1%)"
+    );
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/protocol_fee.rs]: We should also query the trades endpoint (`api/v1/trades?orderUid=<UID>`) and make sure the `executedProtocolFees` returns the expected fee policies and amounts.

File: crates/e2e/tests/e2e/protocol_fee.rs [hunk 7]
```diff
@@ -983,3 +993,384 @@ async fn volume_fee_buy_order_upcoming_future_test(web3: Web3) {
...
+
+    // Verify quote's fee_amount represents pure gas cost (no volume fee)
+    // fee_amount should be: (gas_used * gas_price) / sell_token_price
+    let quote_fee = dai_usdt_quote.quote.fee_amount;
+    tracing::info!(
+        "DAI-USDT quote fee_amount: {} (should only include gas cost, no volume fee)",
+        quote_fee
+    );
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/protocol_fee.rs]: This does not assert anything.

File: crates/e2e/tests/e2e/protocol_fee.rs [hunk 8]
```diff
@@ -983,3 +993,384 @@ async fn volume_fee_buy_order_upcoming_future_test(web3: Web3) {
...
+        .get_native_price(token_dai.address())
+        .await
+        .unwrap();
+    let sell_token_price_f64 = native_price.price;
+    tracing::info!(
+        "DAI native price (wei per token unit): {}",
+        sell_token_price_f64
+    );
+
+    // Calculate expected network fee: (gas * gas_price) / sell_token_price
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/protocol_fee.rs]: There is already an e2e test dedicated to testing the volume fees on quotes. Best to add this logic there.

File: crates/e2e/tests/e2e/protocol_fee.rs [hunk 9]
```diff
@@ -983,3 +993,384 @@ async fn volume_fee_buy_order_upcoming_future_test(web3: Web3) {
...
+        "DAI-USDT executed fee (0% protocol) should match quote fee_amount (pure gas cost) within \
+         ±2%"
+    );
+
+    // executed_fee = network_fee + protocol_fee_in_sell_token
+    // DAI-USDT has 0% protocol fee, so it's our baseline (network fee only)
+    // We use the fee differences to verify protocol fees are applied correctly
+
+    // Test Case 2: USDC-DAI protocol fee component should be ~0.05% of sell_amount
+    // Allow ±2% tolerance for rounding and price conversion
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/protocol_fee.rs]: Having a 2% tolerance while observing a 0.05% difference sounds very overkill. Is this actually needed here?

File: crates/e2e/tests/e2e/protocol_fee.rs [hunk 10]
```diff
@@ -983,3 +993,384 @@ async fn volume_fee_buy_order_upcoming_future_test(web3: Web3) {
...
+        quote_fee >= expected_network_fee * AlloyU256::from(95) / AlloyU256::from(100)
+            && quote_fee <= expected_network_fee * AlloyU256::from(105) / AlloyU256::from(100),
+        "Quote fee_amount should match pure gas cost within ±5% (no volume fee). Expected: {}, \
+         Got: {}",
+        expected_network_fee,
+        quote_fee
+    );
+
+    // Verify executed fee for 0% protocol order matches the quote
+    assert!(
```
Comment (MartinquaXD) [inline crates/e2e/tests/e2e/protocol_fee.rs]: Would probably be good to put this big ass assertion into a function so that you only have to check the logic once.

File: crates/orderbook/src/arguments.rs [hunk 11]
```diff
@@ -166,42 +166,15 @@ pub struct VolumeFeeConfig {
...
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        s.parse::<f64>().map(FeeFactor::try_from)?
-    }
+    /// Custom volume fees for token buckets.
+    /// Format: "factor:token1,token2,..." (e.g.,
+    /// "0:0xA0b86...,0x6B175...,0xdAC17...") Orders where BOTH tokens are
+    /// in the bucket will use the custom fee. Useful for
+    /// stablecoin-to-stablecoin trades or specific token pairs (2-token
+    /// buckets). Multiple buckets can be separated by semicolons.
+    #[clap(long, env, value_delimiter = ';')]
```
Comment (MartinquaXD) [inline crates/orderbook/src/arguments.rs]: AFAIK we always used the default value delimiter (`,`) for seperating the biggest list items. Best to stay consistent here.

File: crates/shared/src/arguments.rs [hunk 12]
```diff
@@ -515,6 +516,86 @@ impl FromStr for ExternalSolver {
...
+pub struct TokenBucketFeeOverride {
+    pub tokens: HashSet<Address>,
+    pub factor: FeeFactor,
+}
+
+impl FromStr for TokenBucketFeeOverride {
+    type Err = anyhow::Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let parts: Vec<&str> = s.split(':').collect();
```
Comment (MartinquaXD) [inline crates/shared/src/arguments.rs]: this can use `split_once()`.

File: crates/shared/src/arguments.rs [hunk 13]
```diff
@@ -515,6 +516,86 @@ impl FromStr for ExternalSolver {
...
+        let tokens: Result<HashSet<Address>, _> = parts[1]
+            .split(',')
+            .map(|token| {
+                token
+                    .parse::<Address>()
+                    .with_context(|| format!("failed to parse token address '{}'", token))
+            })
+            .collect();
+
+        let tokens = tokens?;
```
Comment (MartinquaXD) [inline crates/shared/src/arguments.rs]: nit: no need to have 2 `tokens` variables.
```suggestion
        let tokens = parts[1]
            .split(',')
            .map(|token| {
                token
                    .parse::<Address>()
                    .with_context(|| format!("failed to parse token address '{}'", token))
            })
            .collect::<Result<HashSet<Address>, _>()?;
```

Comment (MartinquaXD) [inline crates/shared/src/arguments.rs]: We should assert that every bucket has at least 2 tokens.

File: crates/autopilot/src/domain/fee/mod.rs [hunk 14]
```diff
@@ -332,30 +341,6 @@ pub enum Policy {
     },
 }
 
-#[derive(Debug, Clone, Copy, PartialEq, Into)]
```
Comment (fafk) [inline crates/autopilot/src/domain/fee/mod.rs:335]: A potentially controversial move: we had a "domain" FeeFactor and a "dto/arguments" FeeFactor that were identical with identical implementation of TryFrom and FromStr. I deleted the "domain" one and just keeping the "dto/record" type. IMHO it was too much Java for no value added.

File: crates/e2e/tests/e2e/protocol_fee.rs (hunk 5 above)
Comment (fafk) [reply crates/e2e/tests/e2e/protocol_fee.rs]: 🤦 I thought "what would I do to get the gas price" and put it there without thinking. Fixed! 🙏

File: crates/e2e/tests/e2e/protocol_fee.rs (hunk 9 above)
Comment (fafk) [reply crates/e2e/tests/e2e/protocol_fee.rs]: The 2% tolerance is on the computed fee, the 0.05% is on the full amount traded, there is a two orders of magnitude difference between those numbers. It is 2% of 0.05%, so the tolerance is actually quite small.

File: crates/shared/src/arguments.rs (hunk 13 above)
Comment (fafk) [reply crates/shared/src/arguments.rs]: This is by design. I can imagine that we want to lower fees on all trades that include e.g. the CoW token.

File: crates/autopilot/src/arguments.rs [hunk 15]
```diff
@@ -593,6 +599,20 @@ pub struct FeePoliciesConfig {
...
     pub upcoming_fee_policies: UpcomingFeePolicies,
+
+    /// Custom volume fees for token buckets.
+    /// Format: "factor:token1;token2;..." (e.g.,
+    /// "0:0xA0b86...;0x6B175...;0xdAC17...") Orders where BOTH tokens are
+    /// in the bucket will use the custom fee. Useful for
+    /// stablecoin-to-stablecoin trades or specific token pairs (2-token
+    /// buckets). Multiple buckets can be separated by commas.
+    #[clap(long, env, value_delimiter = ',')]
+    pub volume_fee_bucket_overrides: Vec<TokenBucketFeeOverride>,
```
Comment (squadgazzz) [inline crates/autopilot/src/arguments.rs]: It seems like storing this as a `HashMap<Factor, HashSet<Address>>` would naturally be safer to avoid duplicate factors.

Comment (fafk) [reply crates/autopilot/src/arguments.rs]: That wouldn't work. You can choose to have 2 token pairs that should have the same factor (say 0%) and having a bucket with 4 of them would mean something else (all combinations vs. just the two pairs).

Comment (squadgazzz) [reply crates/autopilot/src/arguments.rs]: Ah, gotcha, thanks!

File: crates/shared/src/arguments.rs [hunk 16]
```diff
@@ -533,4 +620,59 @@ mod test {
...
+        // Valid inputs with 2 tokens (minimum required)
+        let valid_two_tokens = "0.5:0x0000000000000000000000000000000000000001;\
+                                0x0000000000000000000000000000000000000002";
+        let result = TokenBucketFeeOverride::from_str(valid_two_tokens).unwrap();
+        assert_eq!(result.factor.get(), 0.5);
+        assert_eq!(result.tokens.len(), 2);
+        // Valid inputs with 3 tokens
+        let valid_three_tokens = "0.123:0x0000000000000000000000000000000000000001;\
+                                  0x0000000000000000000000000000000000000002;\
+                                  0x0000000000000000000000000000000000000003";
```
Comment (squadgazzz) [inline crates/shared/src/arguments.rs:669]: Could you also add a test that validates that the `Vec<TokenBucketFeeOverride>` is parsed correctly?

Comment (fafk) [reply crates/shared/src/arguments.rs:669]: That's covered by the e2e test. One would need to introduce a test that deserizalizes a config struct in orderbook/autopilot where it's used to get a unit test coverage.

File: crates/autopilot/src/arguments.rs [hunk 17]
```diff
@@ -593,6 +599,20 @@ pub struct FeePoliciesConfig {
...
+
+    /// Enable volume fees for trades where sell token equals buy token.
+    /// By default, volume fees are NOT applied to same-token trades.
```
Comment (MartinquaXD) [inline crates/autopilot/src/arguments.rs]: > By default, volume fees are NOT applied to same-token trades.

Isn't this currently a required parameter so there is no default?

File: crates/autopilot/src/arguments.rs (hunk 15 above)
Comment (MartinquaXD) [inline crates/autopilot/src/arguments.rs]: These arguments are currently duplicated in the autopilot and orderbook. You can add them to the `shared` arguments to keep them in sync.

File: crates/shared/src/arguments.rs [hunk 18]
```diff
@@ -533,4 +620,59 @@ mod test {
...
     }
+
+    #[test]
+    fn parse_token_bucket_fee_override() {
+        // Valid inputs with 2 tokens (minimum required)
+        let valid_two_tokens = "0.5:0x0000000000000000000000000000000000000001;\
+                                0x0000000000000000000000000000000000000002";
+        let result = TokenBucketFeeOverride::from_str(valid_two_tokens).unwrap();
+        assert_eq!(result.factor.get(), 0.5);
+        assert_eq!(result.tokens.len(), 2);
```
Comment (MartinquaXD) [inline crates/shared/src/arguments.rs:654]: This should also assert that the expected tokens are in the hashset.

File: crates/autopilot/src/domain/fee/policy.rs [hunk 19]
```diff
@@ -84,13 +87,27 @@ impl PriceImprovement {
...
-    pub fn apply(&self, order: &boundary::Order) -> Option<domain::fee::Policy> {
+    pub fn apply(
+        &self,
+        order: &boundary::Order,
+        bucket_overrides: &[TokenBucketFeeOverride],
+        enable_sell_equals_buy_volume_fee: bool,
+    ) -> Option<domain::fee::Policy> {
         match order.metadata.class {
             boundary::OrderClass::Market => None,
             boundary::OrderClass::Liquidity => None,
```
Comment (MartinquaXD) [inline crates/autopilot/src/domain/fee/policy.rs:94]: Not related to this PR itself but is it actually correct that we don't apply the fee to market and liquidity orders? I was under the impression that we moved the decision which fee applies to which order into the config parameters. Any logic that hardcodes the behavior based on order types seems suspicious now.
Should this be an `unreachable!()` instead?

File: crates/shared/src/fee.rs [hunk 20]
```diff
@@ -44,3 +47,30 @@ impl FeeParameters {
         U256::from((fee_in_eth / self.sell_token_price).ceil())
     }
 }
+
+/// Determines the applicable volume fee factor for a token pair, considering
+/// same-token trade configuration, token bucket overrides and default fee
+/// factor.
+pub fn get_applicable_volume_fee_factor(
```
Comment (MartinquaXD) [inline crates/shared/src/fee.rs]: What's the advantage to having this a free function. Currently the autopilot and orderbook both store the necessary parameters differently and pass them into this function. Why not have 1 shared component defined in `shared` that contains all the necessary data, make this a member function, and instantiate 1 of those structs in the orderbook and autopilot?

Comment (m-sz) [reply crates/shared/src/fee.rs]: I agree with Martin here, 'carrying around' the necessary arguments is cumbersome and a single component responsible for the fees will be even better than a free standing function.

File: crates/autopilot/src/arguments.rs (hunk 17 above)
Comment (fafk) [reply crates/autopilot/src/arguments.rs]: It has an implicit default. Default for bool is ArgAction::SetTrue [source](https://docs.rs/clap/latest/clap/_derive/index.html)
```
    /// When encountered, act as if `"true"` was encountered on the command-line
    ///
    /// If no [`default_value`][super::Arg::default_value] is set, it will be `false`.
```

Also proven by the e2e tests which don't set the flag but autopilot & oderbook start just fine.

File: crates/autopilot/src/arguments.rs (hunk 15 above)
Comment (fafk) [reply crates/autopilot/src/arguments.rs]: That way these args can't be where they belong with their friends in [VolumeFeeConfig](https://github.com/cowprotocol/services/blob/custom-volume-fees-per-bucket/crates/orderbook/src/arguments.rs#L152) in the orderbook, but it's still better than having it duplicated. Done.

File: crates/shared/src/fee.rs (hunk 20 above)
Comment (fafk) [reply crates/shared/src/fee.rs]: It took a some refactoring and I had to compromise on the volume fee param which is not really accessible early enough without tearing the relevant segment of the autopilot apart, but this is now done.

File: crates/autopilot/src/domain/fee/policy.rs (hunk 19 above)
Comment (fafk) [reply crates/autopilot/src/domain/fee/policy.rs:94]: Let me investigate this as a follow up, but let's not do unreachable: it might not be hit in staging but then we could see panics in prod. 🥴

**Comment by m-sz** [comment]
The fees situation currently is intertwined as we have the
```rust
// crates/shared/src/fee.rs
pub struct VolumeFeePolicy {
    bucket_overrides: Vec<TokenBucketFeeOverride>,
    default_factor: Option<FeeFactor>,
    enable_sell_equals_buy_volume_fee: bool,
}

//  crates/orderbook/src/arguments.rs
/// Volume-based protocol fee factor to be applied to quotes.
#[derive(clap::Parser, Debug, Clone)]
pub struct VolumeFeeConfig {
    /// This is a decimal value (e.g., 0.0002 for 0.02% or 2 basis points).
    /// The fee is applied to the surplus token (buy token for sell orders,
    /// sell token for buy orders).
    #[clap(...)]
    pub factor: Option<FeeFactor>,

    /// The timestamp from which the volume fee becomes effective.
    #[clap(...)]
    pub effective_from_timestamp: Option<DateTime<Utc>>,
}

// crates/autopilot/src/domain/fee/mod.rs
pub struct ProtocolFee {
    policy: policy::Policy,
    order_class: OrderClass,
}

pub struct ProtocolFees {
    fee_policies: Vec<ProtocolFee>,
    max_partner_fee: FeeFactor,
    upcoming_fee_policies: Option<UpcomingProtocolFees>,
    volume_fee_policy: VolumeFeePolicy,
}

pub struct UpcomingProtocolFees {
    fee_policies: Vec<ProtocolFee>,
    effective_from_timestamp: DateTime<Utc>,
}

// crates/autopilot/src/domain/fee/policy.rs
pub enum Policy {
    Surplus(Surplus),
    PriceImprovement(PriceImprovement),
    Volume(Volume),
}

pub struct Surplus {
    factor: FeeFactor,
    max_volume_factor: FeeFactor,
}

pub struct PriceImprovement {
    factor: FeeFactor,
    max_volume_factor: FeeFactor,
}

pub struct Volume {
    factor: FeeFactor,
}
```

In autopilot, the Volume fee Policy will always use the VolumeFeePolicy to determine token pair volume fee	while always passing override factor
```rust
impl Volume {
    pub fn apply(
        &self,
        order: &boundary::Order,
        volume_fee_policy: &VolumeFeePolicy,
    ) -> Option<domain::fee::Policy> {
        match order.metadata.class {
            boundary::OrderClass::Market => None,
            boundary::OrderClass::Liquidity => None,
            boundary::OrderClass::Limit => {
                // Use shared function to determine applicable volume fee factor
                let factor = volume_fee_policy.get_applicable_volume_fee_factor(
                    order.data.buy_token,
                    order.data.sell_token,
                    Some(self.factor),
                )?;

                Some(domain::fee::Policy::Volume { factor })
            }
        }
    }
}
```

Whereas the orderbook, when constructing `QuoteHandler` will always instantiate the `VolumeFeePolicy` with a default factor as it is known and configurable
```rust
impl QuoteHandler {
    pub fn new(
        order_validator: Arc<dyn OrderValidating>,
        quoter: Arc<dyn OrderQuoting>,
        app_data: Arc<app_data::Registry>,
        volume_fee: Option<VolumeFeeConfig>,
        volume_fee_bucket_overrides: Vec<TokenBucketFeeOverride>,
        enable_sell_equals_buy_volume_fee: bool,
    ) -> Self {
        let volume_fee_policy = VolumeFeePolicy::new(
            volume_fee_bucket_overrides,
            volume_fee.as_ref().and_then(|config| config.factor),
            enable_sell_equals_buy_volume_fee,
        );
...
```

The VolumeFeeConfig can determine an effective-from timestamp. Same as UpcomingProtocolFees in autopilot etc.

It is not for this PR to fix the state of things, but I wanted to point it out, as I lean on the side that the way we specify, calculate and apply fees could be more uniform.
I know orderbook and autopilot have different responsibilities w.r.t. fees but still it is messy to make out and make sure there are no edge cases.

**Comment by fafk** [comment]
>It is not for this PR to fix the state of things, but I wanted to point it out, as I lean on the side that the way we specify, calculate and apply fees could be more uniform.
I know orderbook and autopilot have different responsibilities w.r.t. fees but still it is messy to make out and make sure there are no edge cases.

Agreed, I think it would be ideal to have unified fee configs as much as possible, but it would be a much larger refactor.
